apollo-router/src/services/external.rs

```diff
@@ -96,10 +95,85 @@ pub(crate) struct Externalizable<T> {
     pub(crate) status_code: Option<u16>,
 }
 
+#[buildstructor::buildstructor]
 impl<T> Externalizable<T>
 where
     T: Debug + DeserializeOwned + Serialize + Send + Sync,
 {
+    #[builder(visibility = "pub(crate)")]
+    /// This is the constructor (or builder) to use when constructing a Router
+    /// `Externalizable`.
+    ///
+    fn router_new(
+        stage: PipelineStep,
+        control: Option<Control>,
+        id: Option<String>,
+        headers: Option<HashMap<String, Vec<String>>>,
+        body: Option<T>,
+        context: Option<Context>,
+        status_code: Option<u16>,
+        method: Option<String>,
+        path: Option<String>,
+        sdl: Option<String>,
+    ) -> Self {
+        assert!(matches!(
+            stage,
+            PipelineStep::RouterRequest | PipelineStep::RouterResponse
```

Comment:
I wish this would fail at build time, but it would require two layers of enums... :/